torch/csrc/jit/passes/onnx/list_model_parameters.cpp

```diff
@@ -0,0 +1,164 @@
+#include <torch/csrc/jit/passes/onnx/list_model_parameters.h>
+#include <torch/csrc/jit/jit_log.h>
+#include <torch/csrc/jit/passes/dead_code_elimination.h>
+
+namespace torch {
+namespace jit {
+
+// findSubModuleAttr function chases getAttr chains to locate the submodules.
+// For example:
+// module M {
+//   attributes {
+//     A = <SubModule at ...>
+//   }
+//   ...
+//   %A = prim::GetAttr[name="A"](%self)
+//   ...
+//   %B = prim::GetAttr[name="B"](%A)
+//   ...
+//   %weight = prim::GetAttr[name="scale"](%B)
+//   ...
+std::deque<std::string> findSubModuleAttr(
+    Value* input,
+    std::string& name,
+    Module& attrModule,
+    std::shared_ptr<Graph>& graph) {
+  Node* node = input->node();
+  std::deque<std::string> moduleNames;
+
+  while (!(node->outputs()[0]->type() == graph->inputs()[0]->type())) {
+    if (node->kind() == prim::GetAttr) {
+      moduleNames.push_front(node->s(attr::name));
+      node = node->inputs()[0]->node();
+    } else {
+      return moduleNames;
+    }
+  }
+
+  for (auto& moduleName : moduleNames) {
+    attrModule = attrModule.attr(moduleName).toModule();
+  }
+  return moduleNames;
+}
+
+Value* addParamAsArgument(Function* function, std::string& name, IValue& attr) {
+  auto schema = function->getSchema();
+  auto args = schema.arguments();
+  args.emplace_back(Argument(name, nullptr, c10::nullopt, attr));
+  auto new_schema = FunctionSchema(
+      schema.name(),
+      schema.overload_name(),
+      args,
+      schema.returns(),
+      schema.is_vararg(),
+      schema.is_varret());
+  function->setSchema(new_schema);
+  return function->graph()->addInput(name)->setType(attr.type());
+}
+
+std::vector<IValue> getParamAttributes(
+    std::shared_ptr<Graph>& graph,
+    const Module& module_,
+    Function* function_) {
+  std::vector<IValue> attrValues;
+  auto isEval = !module_.hasattr("training") || !module_.is_training();
+  auto block = graph->block();
+  std::vector<Block*> blocks({block});
+
+  Node* m = *block->nodes().begin();
+  WithInsertPoint guard(m);
+
+  while (!blocks.empty()) {
+    Block* block = blocks.back();
+    blocks.pop_back();
+    for (auto it = block->nodes().begin(); it != block->nodes().end();) {
+      Node* n = *it;
+      it++; // node n can be destroyed
+
+      for (Block* sub_block : n->blocks()) {
+        blocks.emplace_back(sub_block);
+      }
+      if (n->kind() == prim::SetAttr &&
+          n->s(attr::name) == "num_batches_tracked") {
+        n->destroy();
```

Comment:
This is special case handling for batch norm right? Feels like this should move to the part of code getting rid of GetAttr/SetAttr.